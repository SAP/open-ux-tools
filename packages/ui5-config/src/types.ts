export enum OdataVersion {
    v2 = '2',
    v4 = '4'
}
export interface OdataService {
    url?: string;
    client?: string;
    destination?: {
        name: string;
        instance?: string;
    };
    path?: string;
    version: OdataVersion;
    name?: string;
    model?: string;
    metadata?: string;
    annotations?: {
        technicalName: string;
        xml: string;
    };
}

export interface Destination {
    destination: string;
    destinationInstance?: string;
}
export type Backend = Partial<Destination> & {
    path: string;
    url: string;
};

export interface MiddlewareConfig {
    name: string;
    beforeMiddleware?: string;
    afterMiddleware?: string;
    mountPath?: string;
    configuration: {
        port?: number;
        path?: string;
        annotations?: {
            localPath?: string;
            urlPath: string;
        };
        service?: {
            urlBasePath: string;
            name: string;
            metadataXmlPath: string;
            mockdataRootPath?: string;
            generateMockData?: boolean;
        };
        backend?: Backend[];
        ui5?: {
            path: string[];
            url: string;
            version: string | null;
        };
        ignoreCertError?: boolean;
    };
}
