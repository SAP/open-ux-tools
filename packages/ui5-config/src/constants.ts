export const serveStatic = 'fiori-tools-servestatic';
export const fioriToolsProxy = 'fiori-tools-proxy';
