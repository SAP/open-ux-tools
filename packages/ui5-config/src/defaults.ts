export const enum UI5_DEFAULT {
    DEFAULT_UI5_VERSION = '',
    DEFAULT_LOCAL_UI5_VERSION = '1.95.0',
    MIN_UI5_VERSION = '1.60.0',
    MIN_LOCAL_SAPUI5_VERSION = '1.76.0',
    MIN_LOCAL_OPENUI5_VERSION = '1.52.5',
    SAPUI5_CDN = 'https://ui5.sap.com',
    OPENUI5_CDN = 'https://openui5.hana.ondemand.com',
    TYPES_VERSION_SINCE = '1.76.0',
    TYPES_VERSION_BEST_MIN = '1.102.0',
    TYPES_VERSION_PREVIOUS = '1.71.15',
    TYPES_VERSION_BEST = '1.108.0',
    ESM_TYPES_VERSION_SINCE = '1.94.0',
    MANIFEST_VERSION = '1.12.0',
    BASE_COMPONENT = 'sap/ui/core/UIComponent'
}
