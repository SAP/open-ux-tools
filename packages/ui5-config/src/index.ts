export { UI5Config } from './ui5config';
export {
    Configuration,
    CustomTask,
    CustomMiddleware,
    FioriAppReloadConfig,
    FioriToolsProxyConfig,
    FioriToolsProxyConfigBackend,
    FioriToolsProxyConfigUI5,
    AbapApp,
    UI5ProxyConfig,
    UI5ProxyConfigTarget
} from './types';

export { errorCode as yamlErrorCode, YAMLError } from '@sap-ux/yaml';
