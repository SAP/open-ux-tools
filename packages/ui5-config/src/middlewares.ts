import type {
    FioriToolsProxyConfigBackend,
    CustomMiddleware,
    FioriAppReloadConfig,
    FioriToolsProxyConfig,
    MockserverConfig,
    FioriToolsProxyConfigUI5
} from './types';
import type { NodeComment } from '@sap-ux/yaml';

/**
 * Get the configuration for the AppReload middleware.
 *
 * @returns {CustomMiddleware<FioriAppReloadConfig>} the configuration
 */
export function getAppReloadMiddlewareConfig(): CustomMiddleware<FioriAppReloadConfig> {
    return {
        name: 'fiori-tools-appreload',
        afterMiddleware: 'compression',
        configuration: {
            port: 35729,
            path: 'webapp',
            delay: 300
        }
    };
}

/**
 * Get the configuration for the Fiori tools middleware.
 *
 * @param backends configuration of backends
 * @param ui5 UI5 configuration
 * @returns {{config, comments}} configuration and comments
 */
export function getFioriToolsProxyMiddlewareConfig(
    backends?: FioriToolsProxyConfigBackend[],
    ui5?: Partial<FioriToolsProxyConfigUI5>
): {
    config: CustomMiddleware<FioriToolsProxyConfig>;
    comments: NodeComment<CustomMiddleware<FioriToolsProxyConfig>>[];
} {
    const fioriToolsProxy: CustomMiddleware<FioriToolsProxyConfig> = {
        name: 'fiori-tools-proxy',
        afterMiddleware: 'compression',
        configuration: {
            ignoreCertError: false
        }
    };
    const comments: NodeComment<CustomMiddleware<FioriToolsProxyConfig>>[] = [
        {
            path: 'configuration.ignoreCertError',
            comment:
                ' If set to true, certificate errors will be ignored. E.g. self-signed certificates will be accepted'
        }
    ];

    if (backends && backends.length > 0) {
        backends.forEach((element) => {
            element.path = element.path || '/';
        });
        fioriToolsProxy.configuration.backend = backends;
    }

    if (ui5 !== undefined) {
        fioriToolsProxy.configuration['ui5'] = {
            path: ui5.path || ['/resources', '/test-resources'],
            url: ui5.url || 'https://ui5.sap.com'
        };
        if (ui5.version) {
            fioriToolsProxy.configuration['ui5'].version = ui5.version;
        }
        if (ui5.directLoad) {
            fioriToolsProxy.configuration['ui5'].directLoad = true;
        }
    }

    return { config: fioriToolsProxy, comments };
}

export const getMockServerMiddlewareConfig = (path?: string): CustomMiddleware<MockserverConfig> => {
    path = path?.replace(/\/$/, ''); // Mockserver is sensitive to trailing '/'
    return {
        name: 'sap-fe-mockserver',
        beforeMiddleware: 'csp',
        configuration: {
            mountPath: '/',
            services: [
                {
                    urlPath: path || '',
                    metadataPath: './webapp/localService/metadata.xml',
                    mockdataPath: './webapp/localService/data',
                    generateMockData: true
                }
            ],
            annotations: []
        }
    };
};
