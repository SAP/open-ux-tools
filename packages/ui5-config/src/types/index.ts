export * from './ui5yaml';
export * from './middlewares';

export interface AbapApp {
    name: string;
    desription: string;
    package: string;
    transport: string;
}

export interface AbapTarget {
    [key: string]: string | boolean | undefined;
    url?: string;
    client?: string;
    destination?: string;
    scp?: boolean;
}

export interface AbapDeployConfig {
    target: AbapTarget;
    app: AbapApp;
}

export interface FioriAppReloadConfig {
    port: number;
    path: string;
    delay: number;
}
