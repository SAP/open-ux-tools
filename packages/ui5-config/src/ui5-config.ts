import { MiddlewareConfig } from '@sap-ux/open-ux-tools-types';
import { YamlDocument, NodeComment } from '@sap-ux/yaml';

/**
 * Represents a UI5 config file in yaml format (ui5(-*).yaml) with utility functions to manipulate the yaml document.
 *
 * @class UI5Config
 */
export class UI5Config {
    private document: YamlDocument;

    /**
     * Returns a new instance of UI5Config.
     *
     * @static
     * @param {string} serializedYaml - the serialized yaml string
     * @returns {UI5Config} the UI5Config instance
     * @memberof UI5Config
     */
    static async newInstance(serializedYaml: string): Promise<UI5Config> {
        const instance = new UI5Config();
        instance.document = await YamlDocument.newInstance(serializedYaml);
        return instance;
    }

    /**
     * Adds a UI5 Framework entry to the yaml file.
     *
     * @param {string} ui5Version - ui5 version
     * @param {string[]} libraries - a list of libraries
     * @param ui5Theme - ui5 theme
     * @returns {UI5Config} the UI5Config instance
     * @memberof UI5Config
     */
    public addUI5Framework(ui5Version: string, libraries: string[], ui5Theme = 'sap_fiori_3'): UI5Config {
        const libraryObjs = [];
        for (const library of libraries) {
            libraryObjs.push({ name: library });
        }
        // Add theme lib
        ui5Theme = ui5Theme.replace(/_dark$/, ''); // Dark theme versions are provided by base theme lib
        libraryObjs.push({ name: `themelib_${ui5Theme}` });

        this.document.setIn({
            path: 'framework',
            value: { name: 'SAPUI5', version: ui5Version, libraries: libraryObjs }
        });
        return this;
    }

    /**
     * Adds a list of custom middlewares to the config.
     *
     * @param {MiddlewareConfig[]} middlewares - the list of custom middlewares
     * @param {NodeComment<MiddlewareConfig>[]} [comments] - a list of comments
     * @returns {UI5Config} the UI5Config instance
     * @memberof UI5Config
     */
    public addCustomMiddleware(
        middlewares: MiddlewareConfig[],
        comments?: NodeComment<MiddlewareConfig>[],
    ): UI5Config {
        for (const mw of middlewares) {
            this.document.appendTo({ path: 'server.customMiddleware', value: mw, comments });
        }
        return this;
    }

    /**
     * Returns a string representation of the config.
     *
     * @returns {string} the string representation
     * @memberof UI5Config
     */
    public toString(): string {
        return this.document.toString();
    }
}
