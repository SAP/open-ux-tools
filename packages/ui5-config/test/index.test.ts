import type { AbapApp, UI5ProxyConfig } from '../src';
import { UI5Config } from '../src';

describe('UI5Config', () => {
    // values for testing
    const path = '/~testpath~',
        url = 'http://localhost:8080',
        destination = '~destination~',
        destinationInstance = '~destinationInstance~',
        client = '012';

    // object under test
    let ui5Config: UI5Config;
    beforeEach(async () => {
        ui5Config = await UI5Config.newInstance('');
    });

    describe('get/setConfiguration', () => {
        test('get empty configuration', () => {
            expect(ui5Config.getConfiguration()).toMatchObject({});
        });

        test('set first time', () => {
            const config = {
                paths: {
                    webapp: '~/my/webapp'
                }
            };
            ui5Config.setConfiguration(config);
            expect(ui5Config.getConfiguration()).toMatchObject(config);
        });

        test('replace existing', () => {
            ui5Config.setConfiguration({
                propertiesFileSourceEncoding: 'ISO-8859-1',
                paths: {
                    webapp: '~/old/webapp',
                    src: '~/src'
                }
            });
            const config = {
                paths: {
                    webapp: '~/my/webapp'
                }
            };
            ui5Config.setConfiguration(config);
            expect(ui5Config.getConfiguration()).toMatchObject(config);
        });
    });

    describe('setMetadata', () => {
        test('set name and copyright', () => {
            ui5Config.setMetadata({ name: 'test.name', copyright: 'Â©' });
            expect(ui5Config.toString()).toMatchSnapshot();
        });

        test('replace metadata', () => {
            ui5Config.setMetadata({ name: 'replace.me', copyright: 'Should not exist after replace' });
            ui5Config.setMetadata({ name: 'the.replaced.name' });
            expect(ui5Config.toString()).toMatchSnapshot();
        });
    });

    describe('setType', () => {
        test('set type', () => {
            ui5Config.setType('application');
            expect(ui5Config.toString()).toMatchSnapshot();
        });

        test('replace type', () => {
            ui5Config.setType('application');
            ui5Config.setType('library');
            expect(ui5Config.toString()).toMatchSnapshot();
        });
    });

    describe('addUI5Framework', () => {
        test('Minimal set of inputs', () => {
            ui5Config.addUI5Framework('SAPUI5', '1.64.0', []);
            expect(ui5Config.toString()).toMatchSnapshot();
        });

        test('Add with specific theme and additional library', () => {
            ui5Config.addUI5Framework('SAPUI5', '1.64.0', ['sap.m'], 'sap_belize');
            expect(ui5Config.toString()).toMatchSnapshot();
        });

        test('Use a dark theme', () => {
            ui5Config.addUI5Framework('SAPUI5', '1.64.0', ['sap.m'], 'sap_fiori_3_dark');
            expect(ui5Config.toString()).toMatchSnapshot();
        });

        test('Use horizon high dark theme', () => {
            ui5Config.addUI5Framework('SAPUI5', '1.96.0', ['sap.m'], 'sap_horizon_dark');
            expect(ui5Config.toString()).toMatchSnapshot();
        });

        test('Use horizon high contrast white theme', () => {
            ui5Config.addUI5Framework('SAPUI5', '1.96.0', ['sap.m'], 'sap_horizon_hcw');
            expect(ui5Config.toString()).toMatchSnapshot();
        });

        test('Use horizon high contrast black theme', () => {
            ui5Config.addUI5Framework('SAPUI5', '1.96.0', ['sap.m'], 'sap_horizon_hcb');
            expect(ui5Config.toString()).toMatchSnapshot();
        });
    });

    describe('addFioriToolsProxydMiddleware', () => {
        test('add without backend or UI5', () => {
            ui5Config.addFioriToolsProxydMiddleware({});
            expect(ui5Config.toString()).toMatchSnapshot();
        });

        test('add without backend or but UI5 defaults', () => {
            ui5Config.addFioriToolsProxydMiddleware({ ui5: {} });
            expect(ui5Config.toString()).toMatchSnapshot();
        });

        test('add without backend or but all properties for UI5', () => {
            ui5Config.addFioriToolsProxydMiddleware({
                ui5: {
                    directLoad: true,
                    path: ['/~customResources', '/~other'],
                    url: 'http://~url',
                    version: '1.23.3'
                }
            });
            expect(ui5Config.toString()).toMatchSnapshot();
        });

        test('add commonly configured backend (and UI5 defaults)', () => {
            ui5Config.addFioriToolsProxydMiddleware({
                backend: [{ url, path, destination, destinationInstance }],
                ui5: {}
            });
            expect(ui5Config.toString()).toMatchSnapshot();
        });

        test('add backend with flexible parameters (and UI5 defaults)', () => {
            ui5Config.addFioriToolsProxydMiddleware({
                backend: [{ url, path, pathPrefix: '/~prefix', scp: true }],
                ui5: {}
            });
            expect(ui5Config.toString()).toMatchSnapshot();
        });

        test('add backend without destination (and UI5 defaults)', () => {
            ui5Config.addFioriToolsProxydMiddleware({ backend: [{ url, path, client }], ui5: {} });
            expect(ui5Config.toString()).toMatchSnapshot();
        });

        test('add backend without path or destination (and UI5 defaults)', () => {
            ui5Config.addFioriToolsProxydMiddleware({ backend: [{ url }], ui5: {} });
            expect(ui5Config.toString()).toMatchSnapshot();
        });
    });

    describe('addBackendToFioriToolsProxydMiddleware', () => {
        test('add proxy without out backend first and then call add backend', () => {
            ui5Config.addFioriToolsProxydMiddleware({ ui5: {} });
            ui5Config.addBackendToFioriToolsProxydMiddleware({ url, path });
            expect(ui5Config.toString()).toMatchSnapshot();
        });

        test('try adding backend without a proxy middleware added before', () => {
            ui5Config.addFioriToolsAppReloadMiddleware();
            expect(() => ui5Config.addBackendToFioriToolsProxydMiddleware({ url, path })).toThrowError();
        });
    });

    describe('addUi5ToFioriToolsProxydMiddleware', () => {
        test('add ui5 config to empty tools middleware config', () => {
            ui5Config.addFioriToolsProxydMiddleware({});
            ui5Config.addUi5ToFioriToolsProxydMiddleware({
                path: ['/~customResources', '/~other'],
                url: 'http://~url'
            });
            expect(ui5Config.toString()).toMatchSnapshot();
        });
    });

    describe('addMockServerMiddleware', () => {
        test('add with given path', () => {
            ui5Config.addMockServerMiddleware(path);
            expect(ui5Config.toString()).toMatchSnapshot();
        });

        test('add without path', () => {
            ui5Config.addMockServerMiddleware();
            expect(ui5Config.toString()).toMatchSnapshot();
        });
    });

    test('getAppReloadMiddlewareConfig', () => {
        ui5Config.addFioriToolsAppReloadMiddleware();
        expect(ui5Config.toString()).toMatchSnapshot();
    });

    describe('add/find/update/removeCustomMiddleware', () => {
        const customMiddleware = {
            name: 'custom-middleware',
            afterMiddleware: '~otherMiddleware',
            configuration: {
                ui5: {
                    path: ['/resources', '/test-resources'],
                    url: 'http://ui5.example'
                },
                version: '1.95.1',
                debug: true
            } as UI5ProxyConfig
        };
        test('addCustomMiddleware', () => {
            ui5Config.addCustomMiddleware([customMiddleware]);
            expect(ui5Config.toString()).toMatchSnapshot();
        });

        test('findCustomMiddleware', () => {
            ui5Config.addCustomMiddleware([customMiddleware]);
            const found = ui5Config.findCustomMiddleware(customMiddleware.name);
            expect(found).toMatchObject(customMiddleware);
        });

        test('updateMiddleware that did not exist, should add it', () => {
            ui5Config.updateCustomMiddleware(customMiddleware);
            expect(ui5Config.toString()).toMatchSnapshot();
        });

        test('updateMiddleware existing middleware', () => {
            const middlewareUpdate = {
                name: 'custom-middleware',
                afterMiddleware: '~newMiddleware',
                configuration: {
                    newValue: {
                        should: 'overwrite existing'
                    }
                }
            };
            ui5Config.addCustomMiddleware([customMiddleware]);
            ui5Config.updateCustomMiddleware(middlewareUpdate);
            expect(ui5Config.toString()).toMatchSnapshot();
        });

        test('removeMiddleware', () => {
            ui5Config.addCustomMiddleware([customMiddleware]);
            ui5Config.removeCustomMiddleware(customMiddleware.name);
            expect(ui5Config.toString()).toMatchSnapshot();
        });
    });

    describe('add/find/removeCustomTask', () => {
        const customTask = {
            name: 'ui5-task-zipper',
            afterTask: 'generateCachebusterInfo',
            configuration: {
                archiveName: 'my-archive'
            }
        };

        test('addCustomTask', () => {
            ui5Config.addCustomTasks([customTask]);
            expect(ui5Config.toString()).toMatchSnapshot();
        });

        test('findCustomTask', () => {
            const notFound = ui5Config.findCustomTask(customTask.name);
            expect(notFound).toBeUndefined();
            ui5Config.addCustomTasks([customTask]);
            const found = ui5Config.findCustomTask(customTask.name);
            expect(found).toMatchObject(customTask);
            ui5Config.removeCustomTask(customTask.name);
            const removed = ui5Config.findCustomTask(customTask.name);
            expect(removed).toBeUndefined();
        });

        test('removeCustomTask', () => {
            ui5Config.addCustomTasks([customTask]);
            ui5Config.removeCustomTask(customTask.name);
            expect(ui5Config.toString()).toMatchSnapshot();
        });
    });

    describe('addAbapDeployTask', () => {
        const app: AbapApp = {
            name: '~name',
            desription: '~description',
            'package': '~package',
            transport: '~transport'
        };

        test('local settings', () => {
            ui5Config.addAbapDeployTask({ url, client }, app);
            expect(ui5Config.toString()).toMatchSnapshot();
        });

        test('AppStudio + Steampunk settings', () => {
            ui5Config.addAbapDeployTask(
                {
                    destination,
                    scp: true
                },
                app
            );
            expect(ui5Config.toString()).toMatchSnapshot();
        });
    });
});
