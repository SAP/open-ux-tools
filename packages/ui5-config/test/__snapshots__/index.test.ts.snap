// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI5Config add/find/update/removeCustomMiddleware addCustomMiddleware 1`] = `
"server:
  customMiddleware:
    - name: custom-middleware
      afterMiddleware: ~otherMiddleware
      configuration:
        ui5:
          path:
            - /resources
            - /test-resources
          url: http://ui5.example
        version: 1.95.1
        debug: true
"
`;

exports[`UI5Config add/find/update/removeCustomMiddleware removeMiddleware 1`] = `
"server:
  customMiddleware: []
"
`;

exports[`UI5Config add/find/removeCustomTask addCustomTask 1`] = `
"builder:
  customTasks:
    - name: ui5-task-zipper
      afterTask: generateCachebusterInfo
      configuration:
        archiveName: my-archive
"
`;

exports[`UI5Config add/find/removeCustomTask removeCustomTask 1`] = `
"builder:
  customTasks: []
"
`;

exports[`UI5Config add/find/update/removeCustomMiddleware updateMiddleware existing middleware 1`] = `
"server:
  customMiddleware:
    - name: custom-middleware
      afterMiddleware: ~newMiddleware
      configuration:
        newValue:
          should: overwrite existing
"
`;

exports[`UI5Config add/find/update/removeCustomMiddleware updateMiddleware that did not exist, should add it 1`] = `
"server:
  customMiddleware:
    - name: custom-middleware
      afterMiddleware: ~otherMiddleware
      configuration:
        ui5:
          path:
            - /resources
            - /test-resources
          url: http://ui5.example
        version: 1.95.1
        debug: true
"
`;

exports[`UI5Config addAbapDeployTask AppStudio + Steampunk settings 1`] = `
"builder:
  resources:
    - excludes:
        - /test/**
        - /localService/**
  customTasks:
    - name: deploy-to-abap
      afterTask: generateCachebusterInfo
      configuration:
        target:
          destination: ~destination~
          scp: true
        app:
          name: ~name
          desription: ~description
          package: ~package
          transport: ~transport
"
`;

exports[`UI5Config addAbapDeployTask local settings 1`] = `
"builder:
  resources:
    - excludes:
        - /test/**
        - /localService/**
  customTasks:
    - name: deploy-to-abap
      afterTask: generateCachebusterInfo
      configuration:
        target:
          url: http://localhost:8080
          client: '012'
        app:
          name: ~name
          desription: ~description
          package: ~package
          transport: ~transport
"
`;

exports[`UI5Config addBackendToFioriToolsProxydMiddleware add proxy without out backend first and then call add backend 1`] = `
"server:
  customMiddleware:
    - name: fiori-tools-proxy
      afterMiddleware: compression
      configuration:
        ignoreCertError: false # If set to true, certificate errors will be ignored. E.g. self-signed certificates will be accepted
        ui5:
          path:
            - /resources
            - /test-resources
          url: https://ui5.sap.com
        backend:
          - url: http://localhost:8080
            path: /~testpath~
"
`;

exports[`UI5Config addFioriToolsProxydMiddleware add backend with flexible parameters (and UI5 defaults) 1`] = `
"server:
  customMiddleware:
    - name: fiori-tools-proxy
      afterMiddleware: compression
      configuration:
        ignoreCertError: false # If set to true, certificate errors will be ignored. E.g. self-signed certificates will be accepted
        backend:
          - url: http://localhost:8080
            path: /~testpath~
            pathPrefix: /~prefix
            scp: true
        ui5:
          path:
            - /resources
            - /test-resources
          url: https://ui5.sap.com
"
`;

exports[`UI5Config addFioriToolsProxydMiddleware add backend without destination (and UI5 defaults) 1`] = `
"server:
  customMiddleware:
    - name: fiori-tools-proxy
      afterMiddleware: compression
      configuration:
        ignoreCertError: false # If set to true, certificate errors will be ignored. E.g. self-signed certificates will be accepted
        backend:
          - url: http://localhost:8080
            path: /~testpath~
            client: '012'
        ui5:
          path:
            - /resources
            - /test-resources
          url: https://ui5.sap.com
"
`;

exports[`UI5Config addFioriToolsProxydMiddleware add backend without path or destination (and UI5 defaults) 1`] = `
"server:
  customMiddleware:
    - name: fiori-tools-proxy
      afterMiddleware: compression
      configuration:
        ignoreCertError: false # If set to true, certificate errors will be ignored. E.g. self-signed certificates will be accepted
        backend:
          - url: http://localhost:8080
            path: /
        ui5:
          path:
            - /resources
            - /test-resources
          url: https://ui5.sap.com
"
`;

exports[`UI5Config addFioriToolsProxydMiddleware add commonly configured backend (and UI5 defaults) 1`] = `
"server:
  customMiddleware:
    - name: fiori-tools-proxy
      afterMiddleware: compression
      configuration:
        ignoreCertError: false # If set to true, certificate errors will be ignored. E.g. self-signed certificates will be accepted
        backend:
          - url: http://localhost:8080
            path: /~testpath~
            destination: ~destination~
            destinationInstance: ~destinationInstance~
        ui5:
          path:
            - /resources
            - /test-resources
          url: https://ui5.sap.com
"
`;

exports[`UI5Config addFioriToolsProxydMiddleware add without backend or UI5 1`] = `
"server:
  customMiddleware:
    - name: fiori-tools-proxy
      afterMiddleware: compression
      configuration:
        ignoreCertError: false # If set to true, certificate errors will be ignored. E.g. self-signed certificates will be accepted
"
`;

exports[`UI5Config addFioriToolsProxydMiddleware add without backend or but UI5 defaults 1`] = `
"server:
  customMiddleware:
    - name: fiori-tools-proxy
      afterMiddleware: compression
      configuration:
        ignoreCertError: false # If set to true, certificate errors will be ignored. E.g. self-signed certificates will be accepted
        ui5:
          path:
            - /resources
            - /test-resources
          url: https://ui5.sap.com
"
`;

exports[`UI5Config addFioriToolsProxydMiddleware add without backend or but all properties for UI5 1`] = `
"server:
  customMiddleware:
    - name: fiori-tools-proxy
      afterMiddleware: compression
      configuration:
        ignoreCertError: false # If set to true, certificate errors will be ignored. E.g. self-signed certificates will be accepted
        ui5:
          path:
            - /~customResources
            - /~other
          url: http://~url
          version: 1.23.3
          directLoad: true
"
`;

exports[`UI5Config addMockServerMiddleware add with given path 1`] = `
"server:
  customMiddleware:
    - name: sap-fe-mockserver
      beforeMiddleware: csp
      configuration:
        mountPath: /
        services:
          - urlPath: /~testpath~
            metadataPath: ./webapp/localService/metadata.xml
            mockdataPath: ./webapp/localService/data
            generateMockData: true
        annotations: []
"
`;

exports[`UI5Config addMockServerMiddleware add without path 1`] = `
"server:
  customMiddleware:
    - name: sap-fe-mockserver
      beforeMiddleware: csp
      configuration:
        mountPath: /
        services:
          - urlPath: ''
            metadataPath: ./webapp/localService/metadata.xml
            mockdataPath: ./webapp/localService/data
            generateMockData: true
        annotations: []
"
`;

exports[`UI5Config addUI5Framework Add with specific theme and additional library 1`] = `
"framework:
  name: SAPUI5
  version: 1.64.0
  libraries:
    - name: sap.m
    - name: themelib_sap_belize
"
`;

exports[`UI5Config addUI5Framework Minimal set of inputs 1`] = `
"framework:
  name: SAPUI5
  version: 1.64.0
  libraries:
    - name: themelib_sap_fiori_3
"
`;

exports[`UI5Config addUI5Framework Use a dark theme 1`] = `
"framework:
  name: SAPUI5
  version: 1.64.0
  libraries:
    - name: sap.m
    - name: themelib_sap_fiori_3
"
`;

exports[`UI5Config addUI5Framework Use horizon high contrast black theme 1`] = `
"framework:
  name: SAPUI5
  version: 1.96.0
  libraries:
    - name: sap.m
    - name: themelib_sap_horizon
"
`;

exports[`UI5Config addUI5Framework Use horizon high contrast white theme 1`] = `
"framework:
  name: SAPUI5
  version: 1.96.0
  libraries:
    - name: sap.m
    - name: themelib_sap_horizon
"
`;

exports[`UI5Config addUI5Framework Use horizon high dark theme 1`] = `
"framework:
  name: SAPUI5
  version: 1.96.0
  libraries:
    - name: sap.m
    - name: themelib_sap_horizon
"
`;

exports[`UI5Config addUi5ToFioriToolsProxydMiddleware add ui5 config to empty tools middleware config 1`] = `
"server:
  customMiddleware:
    - name: fiori-tools-proxy
      afterMiddleware: compression
      configuration:
        ignoreCertError: false # If set to true, certificate errors will be ignored. E.g. self-signed certificates will be accepted
        ui5:
          - path:
              - /~customResources
              - /~other
            url: http://~url
"
`;

exports[`UI5Config getAppReloadMiddlewareConfig 1`] = `
"server:
  customMiddleware:
    - name: fiori-tools-appreload
      afterMiddleware: compression
      configuration:
        port: 35729
        path: webapp
        delay: 300
"
`;

exports[`UI5Config setMetadata replace metadata 1`] = `
"metadata:
  name: the.replaced.name
"
`;

exports[`UI5Config setMetadata set name and copyright 1`] = `
"metadata:
  name: test.name
  copyright: Â©
"
`;

exports[`UI5Config setType replace type 1`] = `
"type: library
"
`;

exports[`UI5Config setType set type 1`] = `
"type: application
"
`;
